File: src/main/java/me/syari/ss/discord/api/exceptions/ErrorResponseException.java
package me.syari.ss.discord.api.exceptions;

import me.syari.ss.discord.api.requests.ErrorResponse;
import me.syari.ss.discord.api.requests.Response;
import me.syari.ss.discord.api.utils.data.DataObject;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

public class ErrorResponseException extends RuntimeException {
    private ErrorResponseException(Response response, int code, String meaning) {
        super(code + ": " + meaning);
        if (response != null && response.getException() != null) {
            initCause(response.getException());
        }
    }

    @Contract("_, _ -> new")
    public static @NotNull ErrorResponseException create(@NotNull ErrorResponse errorResponse, @NotNull Response response) {
        Optional<DataObject> optObject = response.optObject();
        String meaning = errorResponse.getMeaning();
        int code = errorResponse.getCode();
        if (response.isError() && response.getException() != null) {
            code = response.code;
            meaning = response.getException().getClass().getName();
        } else if (optObject.isPresent()) {
            DataObject object = optObject.get();
            boolean isNullCode = object.isNull("code");
            boolean isNullMessage = object.isNull("message");
            if (!isNullCode || !isNullMessage) {
                if (!isNullCode) {
                    code = object.getInt("code");
                }
                if (!isNullMessage) {
                    meaning = object.getString("message");
                }
            } else {
                code = response.code;
                meaning = object.toString();
            }
        } else {
            code = response.code;
            meaning = response.getString();
        }

        return new ErrorResponseException(response, code, meaning);
    }
}

File: src/main/java/me/syari/ss/discord/api/exceptions/ErrorResponseException.kt
package me.syari.ss.discord.api.exceptions

import me.syari.ss.discord.api.requests.ErrorResponse
import me.syari.ss.discord.api.requests.Response

class ErrorResponseException private constructor(
    response: Response?, code: Int, meaning: String
): RuntimeException("$code: $meaning") {
    init {
        if (response?.exception != null) {
            initCause(response.exception)
        }
    }

    companion object {
        fun create(
            errorResponse: ErrorResponse, response: Response
        ): ErrorResponseException {
            val dataObject = response.optObject()
            var meaning = errorResponse.meaning
            var code = errorResponse.code
            if (response.isError && response.exception != null) {
                code = response.code
                meaning = response.exception?.javaClass?.name ?: "null"
            } else if (dataObject != null) {
                val nullableCode = dataObject.getInt("code")
                val nullableMessage = dataObject.getString("message")
                if (nullableCode != null || nullableMessage != null) {
                    if (nullableCode != null) code = nullableCode
                    if (nullableMessage != null) meaning = nullableMessage
                } else {
                    code = response.code
                    meaning = dataObject.toString()
                }
            } else {
                code = response.code
                meaning = response.string
            }
            return ErrorResponseException(response, code, meaning)
        }
    }

}